### PR TITLE
Ensure that events are correctly updated when they are edited.

### PR DESCRIPTION
**Pull request by richvdh**
Make sure that we show the correct details of edited events.

Notes to the reviewer:
 * There are a few refactors here before the main commit. It's probably going to be easier to review them commit-by-commit.
 * Currently this is failing the test coverage gate; this is mostly due to the changes to the calls to `verifyEvent` which aren't well covered. I've spent quite a lot of time adding tests for this change, and I really can't afford to get even deeper into that rabbithole. I believe the changes to `verifyEvent` are pretty trivial.

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Ensure that events are correctly updated when they are edited. ([\#9789](https://github.com/matrix-org/matrix-react-sdk/pull/9789)).

### CODE REVIEW

File: test/components/views/rooms/EventTile-test.tsx [hunk 1]
```diff
@@ -67,7 +71,7 @@ describe("EventTile", () => {
         stubClient();
         client = MatrixClientPeg.get();
 
-        room = new Room(ROOM_ID, client, client.getUserId(), {
+        room = new Room(ROOM_ID, client, client.getUserId()!, {
```
Comment (richvdh) [inline test/components/views/rooms/EventTile-test.tsx:74]: a couple of drive-by strict typecheck fixes while I was in the area.

File: src/components/views/rooms/EventTile.tsx [hunk 2]
```diff
@@ -470,7 +470,11 @@ export class UnwrappedEventTile extends React.Component<EventTileProps, IState>
         this.threadState?.off(NotificationStateEvents.Update, this.onThreadStateUpdate);
     }
 
-    public componentDidUpdate(prevProps: Readonly<EventTileProps>) {
+    public componentDidUpdate(prevProps: Readonly<EventTileProps>, prevState: Readonly<IState>) {
+        // If the verification state changed, the height might have changed
+        if (prevState.verified !== this.state.verified) {
+            this.props.onHeightChanged();
+        }
```
Comment (richvdh) [inline src/components/views/rooms/EventTile.tsx:479]: This is (imho) a simplification (and in line with react's recommendations), but also turns out to be needed to keep some other tests passing, to ensure we don't call onHeightChanged if the verified state stays at `null`.

**Review by dbkr — APPROVED** [review]
